### PR TITLE
CLCT RX and related changes from Marco

### PR DESCRIPTION
**Pull request by slava77**
I dropped most of the re-indentation from the original changes made by Marco.
I've asked for indents to be more reasonable, but, in retrospect, I needed this only for the new code.

### CODE REVIEW

File: emuDCS/PeripheralApps/src/common/ChamberUtilities.cc [hunk 1]
```diff
@@ -462,54 +464,2132 @@ ChamberUtilities::ChamberUtilities(){
...
+      thisTMB->WriteRegister(seq_l1a_adr);
+    }
+
+    int ChamberUtilities::GetTMBInternalL1ADelay() {
+      thisTMB->ReadRegister(seq_l1a_adr);
+      //return thisTMB->GetInternalL1aDelay();
+      return thisTMB->GetInternalL1a();
+    }
+
+    void ChamberUtilities::ResyncDCFEBs() {
```
Comment (slava77) [inline emuDCS/PeripheralApps/src/common/ChamberUtilities.cc]: move implementation  to DAQMB at some point

File: emuDCS/PeripheralApps/src/common/ChamberUtilities.cc [hunk 2]
```diff
@@ -462,54 +464,2132 @@ ChamberUtilities::ChamberUtilities(){
...
+	    addr = 0x3010/*DCFEB_REPROGRAM*/ | (slot_number << 19);
+	    data = 0x0002;
+	  }
+	  thisDMB->getCrate()->vmeController()->vme_controller(irdwr, addr, &data, rcv);
+	}
+	std::cout << " Calling Resync " << std::endl;
+      } // ODMB only
+    }
+
+    void ChamberUtilities::SetDCFEBsPipelineDepth(int depth) {
```
Comment (slava77) [inline emuDCS/PeripheralApps/src/common/ChamberUtilities.cc]: move implementation  to DAQMB at some point

File: emuDCS/PeripheralApps/src/common/ChamberUtilities.cc [hunk 3]
```diff
@@ -462,54 +464,2132 @@ ChamberUtilities::ChamberUtilities(){
...
+      if(thisDMB->getCrate()) {
+	thisDMB->getCrate()->ccb()->l1aReset(); // need to do this after restarting DCFEB pipelines
+	usleep(1000);
+	//resyncDCFEBs(thisDMB->getCrate()); // TODO: remove once firmware takes care of it
+	ResyncDCFEBs();
+	usleep(1000);
+      }
+    }
+
+    void ChamberUtilities::CFEBTiming_ConfigureODMB() {
```
Comment (slava77) [inline emuDCS/PeripheralApps/src/common/ChamberUtilities.cc]: move implementation  to DAQMB at some point

File: emuDCS/PeripheralApps/src/common/ChamberUtilities.cc [hunk 4]
```diff
@@ -462,54 +464,2132 @@ ChamberUtilities::ChamberUtilities(){
...
+
+	// Temporary fix to explicitly set crate id in ODMB
+	irdwr = 3;
+	addr = 0x4020/*ODMB_CRATEID*/ | slot_number << 19;
+	data = thisTMB->getCrate()->CrateID();
+	thisDMB->getTheController()->vme_controller(irdwr, addr, &data, rcv);
+      }
+    }
+
+    void ChamberUtilities::SetODMBPedestalMode() {
```
Comment (slava77) [inline emuDCS/PeripheralApps/src/common/ChamberUtilities.cc]: move implementation  to DAQMB at some point

File: emuDCS/PeripheralApps/src/common/ChamberUtilities.cc [hunk 5]
```diff
@@ -462,54 +464,2132 @@ ChamberUtilities::ChamberUtilities(){
...
+      data = 0x000f;
+      thisDMB->getCrate()->vmeController()->vme_controller(irdwr, addr, &data, rcv);
+    }
+    //
+    int ChamberUtilities::GetExtTrigDelay() {
+      unsigned csrb5 = thisCCB_->ReadRegister(CCB::CSRB5);
+      return (csrb5 >> 8) & 0xff;
+    }
+
+    void ChamberUtilities::Clear_ODMB_FIFO() {
```
Comment (slava77) [inline emuDCS/PeripheralApps/src/common/ChamberUtilities.cc]: move implementation  to DAQMB at some point

File: emuDCS/PeripheralApps/src/common/ChamberUtilities.cc [hunk 6]
```diff
@@ -462,54 +464,2132 @@ ChamberUtilities::ChamberUtilities(){
...
+      int n_words = thisDMB->ReadRegister(0x550C);
+      for(int i=0; i<n_words; ++i) {
+	thisDMB->ReadRegister(0x5500);
+      }
+    }
+
+    void ChamberUtilities::Print_ODMB_FIFO() {
```
Comment (slava77) [inline emuDCS/PeripheralApps/src/common/ChamberUtilities.cc]: move implementation  to DAQMB at some point

File: emuDCS/PeripheralApps/src/common/EmuPeripheralCrateConfig.cc [hunk 7]
```diff
@@ -5055,68 +5084,88 @@ void EmuPeripheralCrateConfig::ChamberTests(xgi::Input * in, xgi::Output * out )
...
+  //
+  *out << cgicc::table();
+  //
+  *out << cgicc::tr();
+  //
+  *out << cgicc::tr();
+  //
+  *out << cgicc::td();
   //
   for(int i=0;i<5;i++) {
```
Comment (austinschneider) [inline emuDCS/PeripheralApps/src/common/EmuPeripheralCrateConfig.cc]: Why is this printing only 5?

Comment (austinschneider) [inline emuDCS/PeripheralApps/src/common/EmuPeripheralCrateConfig.cc]: ChamberUtilities.h:  `inline int  GetCFEBrxPhaseTest(int CFEB) { return CFEBrxPhase_[CFEB] ; }`
the CFEBrxPhase_ is size 7, makes sense to print more if it's OTMB
... fixing it

**Comment by austinschneider** [comment]
(actually notes by @slava77 )
Summary of changes, from visual inspection of the EmuPeripheralCrateConfig:
- CrateConfiguration Muon timing section (at the bottom) doesn't have the "Find CFEB rx for crate" button
- ExpertToolsPage
  -  "CFEB rx" button in "Determine online synchronization parameters" is removed
  - "Scan ME1/1s on All Crates" is added at the bottom in the "DCFEB RX Scan on ME1/1s over All Crates" section
  - "Scan ME1/1s on All Crates" is added in the "CFEB RX Scan on non-ME1/1s over All Crates" section
- in ChamberTests
  - "Team A tests" section: the "ME1/1b Scan CFEB rx phase"  and "ME1/1a Scan CFEB rx phase" buttons  are removed
  - "RX Delay     CFEB Clock Phase    CFEB    Pattern Type    Half-Strip" choices and corresponding test is added
    screenshot_o
    screenshot_a
    After running some random test:
    screenshot_aftertest
    ... after fixup in c15ca18 ... 
    teama_aftertest

**Comment by slava77** [comment]
Some notes after running the scan at 904

clctscan_result

```
Chamber-Crate Phases ME-1/1/08 output:
*************************************
          Test CFEB Timing           
*************************************
     tmb_internal_l1a:    0
 clct_pattern_trig_en:    1
     clct_ext_trig_en:    1
       tmb_allow_clct:    1
hs_pretrig_hit_thresh:    5
     min_hits_pattern:    1
 ignore_ccb_startstop:    0
     layer_trigger_en:    0
            fifo_mode:    1
           fifo_tbins:   30
         fifo_pretrig:    5
     fifo_no_hits_raw:    0
   ccb_ext_trig_delay:    0
        tmb_l1a_delay:    0
       cfeb_rx_posneg:    0
  cfeb_rx_clock_delay:    0
     cfeb_clock_phase:    0
     tmb_internal_l1a:    0
 clct_pattern_trig_en:    1
     clct_ext_trig_en:    1
       tmb_allow_clct:    1
hs_pretrig_hit_thresh:    5
     min_hits_pattern:    1
 ignore_ccb_startstop:    0
     layer_trigger_en:    0
            fifo_mode:    1
           fifo_tbins:   30
         fifo_pretrig:    5
     fifo_no_hits_raw:    0
   ccb_ext_trig_delay:    0
        tmb_l1a_delay:    0
       cfeb_rx_posneg:    0
  cfeb_rx_clock_delay:    0
     cfeb_clock_phase:    0
----------------------
CLCT0.Valid      = 0x0
CLCT0.Nhits      = 0x0
CLCT0.pattern    = 0x0
CLCT0.Key HStrip = 0
CLCT0.sync err   = 0x1


Errors per cfeb: 
CFEB 0: 9
CFEB 1: 1
CFEB 2: 5
CFEB 3: 17
CFEB 4: 21
CFEB 5: 16
CFEB 6: 10

CFEB Rx Windows: 
Posneg: 0
   RX |    0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16   17   18   19   20   21   22   23   24
    0 |    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    4    0    0    0    0
    1 |    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    2 |    0    0    0    0    0    0    0    0    0    5    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    3 |    0    0    0    0    0    0    0    0    0    7    0    0    0    6    0    0    0    0    0    0    0    0    0    0    0
    4 |    0    0    0    0    0    0    0    0    0   11    0    0    0    6    0    0    0    0    0    0    0    0    0    0    0
    5 |    0    0    0    0    0    0    0    0    0    6    0    0    0    0    0    0    0    0    0    6    0    0    0    0    0
    6 |    0    0    0    0    0    0    0    0    0    7    0    0    0    1    0    0    0    0    0    0    0    0    0    0    0
Posneg: 1
   RX |    0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16   17   18   19   20   21   22   23   24
    0 |    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    5    0    0    0    0
    1 |    0    0    0    0    0    0    0    0    0    0    0    0    0    1    0    0    0    0    0    0    0    0    0    0    0
    2 |    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    3 |    0    0    0    0    0    0    0    0    0    0    0    0    0    4    0    0    0    0    0    0    0    0    0    0    0
    4 |    0    0    0    0    0    0    0    0    0    0    0    0    0    4    0    0    0    0    0    0    0    0    0    0    0
    5 |    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    2    0    0    2    0    0
    6 |    0    0    0    0    0    0    0    0    0    0    0    0    0    2    0    0    0    0    0    0    0    0    0    0    0
--------------------------------

ERROR: No window edge found!
ERROR: No window edge found!
Posneg: 0
 CFEB |Best RX   Ind. RX
    0 |      2         8
ERROR: No window edge found!
    1 |      2        -1
ERROR: No window edge found!
    2 |      2        22
    3 |      2        24
    4 |      2        24
    5 |      2         2
    6 |      2        24

Posneg: 1
 CFEB |Best RX   Ind. RX
    0 |      4         8
    1 |      4         1
ERROR: No window edge found!
    2 |      4        -1
ERROR: No window edge found!
    3 |      4         1
    4 |      5         1
    5 |      5         8
    6 |      5         1

Using parameters: 

DCFEB |   RX Posneg
    0 |    4      1
    1 |    4      1
    2 |    4      1
    3 |    4      1
    4 |    5      1
    5 |    5      1
    6 |    5      1

Reverting back to original cfeb delay phase values...



Total run time: 163 s

```

**Comment by slava77** [comment]
and then one more run immediately after:

clctscan_result2

```
*************************************
          Test CFEB Timing           
*************************************
     tmb_internal_l1a:    0
 clct_pattern_trig_en:    1
     clct_ext_trig_en:    1
       tmb_allow_clct:    1
hs_pretrig_hit_thresh:    5
     min_hits_pattern:    1
 ignore_ccb_startstop:    0
     layer_trigger_en:    0
            fifo_mode:    1
           fifo_tbins:   30
         fifo_pretrig:    5
     fifo_no_hits_raw:    0
   ccb_ext_trig_delay:    0
        tmb_l1a_delay:    0
       cfeb_rx_posneg:    0
  cfeb_rx_clock_delay:    0
     cfeb_clock_phase:    0
     tmb_internal_l1a:    0
 clct_pattern_trig_en:    1
     clct_ext_trig_en:    1
       tmb_allow_clct:    1
hs_pretrig_hit_thresh:    5
     min_hits_pattern:    1
 ignore_ccb_startstop:    0
     layer_trigger_en:    0
            fifo_mode:    1
           fifo_tbins:   30
         fifo_pretrig:    5
     fifo_no_hits_raw:    0
   ccb_ext_trig_delay:    0
        tmb_l1a_delay:    0
       cfeb_rx_posneg:    0
  cfeb_rx_clock_delay:    0
     cfeb_clock_phase:    0
----------------------
CLCT0.Valid      = 0x0
CLCT0.Nhits      = 0x0
CLCT0.pattern    = 0x0
CLCT0.Key HStrip = 0
CLCT0.sync err   = 0x1


Errors per cfeb: 
CFEB 0: 13
CFEB 1: 10
CFEB 2: 9
CFEB 3: 2
CFEB 4: 16
CFEB 5: 16
CFEB 6: 11

CFEB Rx Windows: 
Posneg: 0
   RX |    0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16   17   18   19   20   21   22   23   24
    0 |    0    0    0    0    0    0    0    0    0    1    0    0    0    0    0    0    0    0    0    2    0    0    0    0    0
    1 |    0    0    0    0    0    0    0    0    0    0    0    0    3    0    0    0    0    0    0    0    0    0    0    0    0
    2 |    0    0    0    0    0    0    0    0    0    5    0    0    1    0    0    0    0    0    0    0    0    0    0    0    0
    3 |    0    0    0    0    0    0    0    0    0    2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    4 |    0    0    0    0    0    0    0    0    0    4    0    3    0    0    0    0    0    0    0    0    0    0    0    0    0
    5 |    0    0    0    0    0    0    0    0    0    3    0    0    0    0    0    0    0    7    0    0    0    0    0    0    0
    6 |    0    0    0    0    0    0    0    0    0    3    0    3    0    0    0    0    0    0    0    0    0    0    0    0    0
Posneg: 1
   RX |    0    1    2    3    4    5    6    7    8    9   10   11   12   13   14   15   16   17   18   19   20   21   22   23   24
    0 |    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    8    0    0    2    0    0
    1 |    0    0    0    0    0    0    0    0    0    0    0    0    6    0    0    0    0    0    0    0    0    0    1    0    0
    2 |    0    0    0    0    0    0    0    0    0    0    0    0    3    0    0    0    0    0    0    0    0    0    0    0    0
    3 |    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    4 |    0    0    0    0    0    0    0    0    0    0    0    5    0    0    0    0    0    0    0    0    0    0    4    0    0
    5 |    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    2    0    0    0    0    4    0    0
    6 |    0    0    0    0    0    0    0    0    0    0    0    5    0    0    0    0    0    0    0    0    0    0    0    0    0
--------------------------------

ERROR: No window edge found!
Posneg: 0
 CFEB |Best RX   Ind. RX
    0 |      2         2
    1 |      2         0
    2 |      2        23
    3 |      2        22
    4 |      1        23
    5 |      1         1
    6 |      1        23

Posneg: 1
 CFEB |Best RX   Ind. RX
    0 |      5         8
    1 |      5         5
    2 |      5         0
ERROR: No window edge found!
    3 |      5        -1
ERROR: No window edge found!
    4 |      4         4
    5 |      4         7
    6 |      4        24

Using parameters: 

DCFEB |   RX Posneg
    0 |    2      0
    1 |    2      0
    2 |    2      0
    3 |    2      0
    4 |    1      0
    5 |    1      0
    6 |    1      0

Reverting back to original cfeb delay phase values...



Total run time: 163 s


```

After the scan the link status is mostly in error

status_afterclctscan

**Pull request merged**